agent/containers_internal_test.go

```diff
@@ -0,0 +1,206 @@
+			// ory/dockertest pre-pends a forward slash to the container name.
+			assert.Equal(t, strings.TrimPrefix(ct.Container.Name, "/"), foundContainer.FriendlyName)
+			// ory/dockertest returns the sha256 digest of the image.
+			assert.Equal(t, "busybox:latest", foundContainer.Image)
+			assert.Equal(t, ct.Container.Config.Labels, foundContainer.Labels)
+			assert.True(t, foundContainer.Running)
+			assert.Equal(t, "running", foundContainer.Status)
+			assert.Len(t, foundContainer.Ports, 0)
+			assert.Len(t, foundContainer.Volumes, 0)
+			break
+		}
+	}
+	assert.True(t, found, "Expected to find container with label 'com.coder.test=%s'", testLabelValue)
+}
+
+// TestContainersHandler tests the containersHandler.getContainers method using
+// a mock implementation. It specifically tests caching behavior.
+func TestContainersHandler(t *testing.T) {
+	t.Parallel()
+
+	t.Run("list", func(t *testing.T) {
+		t.Parallel()
+
+		fakeCt := fakeContainer(t)
+		fakeCt2 := fakeContainer(t)
+
+		// Each test case is called multiple times to ensure idempotency
+		for _, tc := range []struct {
+			name string
+			// data to be stored in the handler
+			cacheData []codersdk.WorkspaceAgentContainer
+			// duration of cache
+			cacheDur time.Duration
+			// relative age of the cached data
+			cacheAge time.Duration
+			// function to set up expectations for the mock
+			setupMock func(*MockContainerLister)
+			// expected result
+			expected []codersdk.WorkspaceAgentContainer
+			// expected error
+			expectedErr string
+		}{
+			{
+				name: "no cache",
+				setupMock: func(mcl *MockContainerLister) {
+					mcl.EXPECT().List(gomock.Any()).Return([]codersdk.WorkspaceAgentContainer{fakeCt}, nil).AnyTimes()
+				},
+				expected: []codersdk.WorkspaceAgentContainer{fakeCt},
+			},
+			{
+				name:      "no data",
+				cacheData: nil,
+				cacheAge:  2 * time.Second,
+				cacheDur:  time.Second,
+				setupMock: func(mcl *MockContainerLister) {
+					mcl.EXPECT().List(gomock.Any()).Return([]codersdk.WorkspaceAgentContainer{fakeCt}, nil).AnyTimes()
+				},
+				expected: []codersdk.WorkspaceAgentContainer{fakeCt},
+			},
+			{
+				name:      "cached data",
+				cacheAge:  time.Second,
+				cacheData: []codersdk.WorkspaceAgentContainer{fakeCt},
+				cacheDur:  2 * time.Second,
+				expected:  []codersdk.WorkspaceAgentContainer{fakeCt},
+			},
+			{
+				name: "lister error",
+				setupMock: func(mcl *MockContainerLister) {
+					mcl.EXPECT().List(gomock.Any()).Return(nil, assert.AnError).AnyTimes()
+				},
+				expectedErr: assert.AnError.Error(),
+			},
+			{
+				name:      "stale cache",
+				cacheAge:  2 * time.Second,
+				cacheData: []codersdk.WorkspaceAgentContainer{fakeCt},
+				cacheDur:  time.Second,
+				setupMock: func(mcl *MockContainerLister) {
+					mcl.EXPECT().List(gomock.Any()).Return([]codersdk.WorkspaceAgentContainer{fakeCt2}, nil).AnyTimes()
+				},
+				expected: []codersdk.WorkspaceAgentContainer{fakeCt2},
+			},
+		} {
+			tc := tc
```

Comment:
Nit: no longer necessary as of Go v1.21.

Reply:
I know, my linter just complains.